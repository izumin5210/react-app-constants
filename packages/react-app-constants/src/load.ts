/* eslint eqeqeq: ["error", "always", {"null": "ignore"}] */

import { Base64 } from "js-base64";

const metaName = "app-constants";

export function loadConstants<T>(): T {
  if (typeof window !== "undefined" && typeof window.document !== "undefined") {
    return loadConstantsFromMeta();
  }
  return loadConstantsFromEnv();
}

function loadConstantsFromMeta<T>(): T {
  const v = document
    .querySelector(`meta[name="${metaName}"]`)
    ?.getAttribute("content");
  if (v == null) {
    throw new Error(`meta[name="${metaName}"] is not found or has no contents`);
  }
  return JSON.parse(Base64.decode(v)) as T;
}

function loadConstantsFromEnv<T>(): T {
  const entries = Object.entries(process.env)
    .filter(([k, _]) => k.startsWith("REACT_APP_"))
    .map(([k, v]) => [k.replace(/^REACT_APP_/, ""), v]);
  return Object.fromEntries(entries) as T;
}
