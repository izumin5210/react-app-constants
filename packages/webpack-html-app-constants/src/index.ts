/* eslint eqeqeq: ["error", "always", {"null": "ignore"}] */

import dotenv from "dotenv";
import * as fs from "fs";
import { Base64 } from "js-base64";
import { join } from "path";
import webpack from "webpack";
import {
  Overrider,
  rewire as rewireToPowerstrip,
} from "webpack-html-powerstrip";

interface GlobalConfig {
  dir: string;
  fs: Pick<webpack.InputFileSystem, "readFileSync">;
}

export const config: GlobalConfig = {
  dir: fs.realpathSync(process.cwd()),
  fs,
};

export interface Params {
  variations: string[];
  loader?: Loader;
  transformers?: Transformer[];
}

export function rewire(
  config: webpack.Configuration,
  env: string,
  { loader = loadFromDotenv, transformers = [onlyReactApp], variations }: Params
) {
  const overrider = createOverrider(loader, transformers);
  return rewireToPowerstrip(config, env, {
    variations,
    overrider,
  });
}

type Constants = [string, string][];
type Loader = (variation?: string) => Constants;
type Transformer = (consts: Constants, variation?: string) => Constants;

function createOverrider(load: Loader, transformers: Transformer[]): Overrider {
  return (variation?: string) => {
    let entries = load(variation);
    for (const f of transformers) {
      entries = f(entries);
    }
    const cfg = Object.fromEntries(entries);

    const meta = { "app-constants": Base64.encode(JSON.stringify(cfg)) };
    return { meta };
  };
}

const loadFromDotenv: Loader = variation => {
  const path = join(
    config.dir,
    variation == null ? ".env" : `.env.${variation}`
  );
  const cfg = dotenv.parse(config.fs.readFileSync(path)) || {};
  return Object.entries(cfg);
};

const onlyReactApp: Transformer = (consts, _) =>
  consts
    .filter(([k, _]) => k.startsWith("REACT_APP_"))
    .map(([k, v]) => [k.replace(/^REACT_APP_/, ""), v]);
