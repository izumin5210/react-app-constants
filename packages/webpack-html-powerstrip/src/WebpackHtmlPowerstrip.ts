/* eslint eqeqeq: ["error", "always", {"null": "ignore"}] */

import deepmerge from "deepmerge";
import * as HtmlWebpackPlugin from "html-webpack-plugin";
import * as Webpack from "webpack";

export type Overrider = (variant?: string) => HtmlWebpackPlugin.Options;

export interface Params {
  variants: string[];
  overrider: Overrider;
}

export default class WebpackHtmlPowerstrip {
  private variants: string[];
  private overrider: Overrider;

  constructor({ variants, overrider }: Params) {
    this.variants = variants;
    this.overrider = overrider;
  }

  public rewire(
    config: Webpack.Configuration,
    env: string
  ): Webpack.Configuration {
    if (config.plugins == null) {
      return config;
    }

    const basePlugin = this.popHtmlWebpackPlugin(config);
    if (basePlugin == null) {
      return config;
    }

    if (env === "production") {
      for (const variant of this.variants) {
        config.plugins.push(
          this.createHtmlWebpackPlugin(basePlugin, variant)
        );
      }
    } else {
      config.plugins.push(this.createHtmlWebpackPlugin(basePlugin));
    }

    return config;
  }

  private popHtmlWebpackPlugin(
    config: Webpack.Configuration
  ): Webpack.Plugin | null {
    let plugin = null;

    config.plugins = (config.plugins || []).filter(p => {
      if (p.constructor.name !== "HtmlWebpackPlugin") {
        return true;
      }
      plugin = p;
      return false;
    });

    return plugin;
  }

  private createHtmlWebpackPlugin(base: Webpack.Plugin, variant?: string) {
    return this.cloneHtmlWebpackPlugin(base, {
      filename: variant == null ? "index.html" : `index.${variant}.html`,
      ...this.overrider(variant),
    });
  }

  private cloneHtmlWebpackPlugin(
    base: Webpack.Plugin,
    override: HtmlWebpackPlugin.Options
  ) {
    const plugin = Object.assign(
      Object.create(Object.getPrototypeOf(base)),
      base
    );
    plugin.options = deepmerge(plugin.options, override);
    return plugin;
  }
}
