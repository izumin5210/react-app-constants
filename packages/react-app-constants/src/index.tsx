/* eslint eqeqeq: ["error", "always", {"null": "ignore"}] */

import React, { createContext, useContext } from "react";
import { loadConstants } from "./load";

type Context<T> = React.Context<T | null>;
type Provider<T> = React.FC<{ value?: T; loader?: () => T }>;
type HOC<T> = <P>(Wrapped: React.ComponentType<P & T>) => React.FC<P>;

export function createConstants<T>(): {
  ConstantsContext: Context<T>;
  ConstantsProvider: Provider<T>;
  useConstants: () => T;
  withConstants: HOC<T>;
} {
  const ConstantsContext = createContext<T | null>(null);
  const useConstants = createHook(ConstantsContext);

  return {
    ConstantsContext,
    ConstantsProvider: createProvider(ConstantsContext),
    useConstants,
    withConstants: createHOC(useConstants),
  };
}

function createProvider<T>(ctx: Context<T>) {
  const ConstantsProvider: Provider<T> = ({
    children,
    value,
    loader = loadConstants,
  }) => <ctx.Provider value={value || loader()}>{children}</ctx.Provider>;
  return ConstantsProvider;
}

function createHook<T>(ctx: Context<T>): () => T {
  return () => {
    const consts = useContext(ctx);
    if (consts == null) {
      throw Error("constans is not loaded");
    }
    return consts;
  };
}

function createHOC<T>(useConstants: () => T) {
  return function<P>(Wrapped: React.ComponentType<P & T>) {
    const Constants: React.FC<P> = props => (
      <Wrapped {...{ ...props, ...useConstants() }} />
    );
    return Constants;
  };
}
