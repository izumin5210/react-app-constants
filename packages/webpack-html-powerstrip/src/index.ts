import * as webpack from "webpack";
import WebpackHtmlPowerstrip, { Params } from "./WebpackHtmlPowerstrip";

function rewire(
  config: webpack.Configuration,
  env: string,
  params: Params
): webpack.Configuration {
  return new WebpackHtmlPowerstrip(params).rewire(config, env);
}

export { WebpackHtmlPowerstrip, rewire };
