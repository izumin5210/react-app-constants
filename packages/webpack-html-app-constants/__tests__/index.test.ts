import HtmlWebpackPlugin from "html-webpack-plugin";
import { createFsFromVolume, Volume } from "memfs";
import { join } from "path";
import webpack from "webpack";
import { config as globalConfig, rewire } from "../src";

const outputPath = join(__dirname, "..", "dist");

function setup(): {
  fs: webpack.OutputFileSystem;
  createDotenv: (variation: string, pairs: string[]) => void;
  expectToMatchSnapshot: (filename: string) => void;
} {
  const vol = new Volume();
  const fs = createFsFromVolume(vol);

  globalConfig.dir = "/app";
  globalConfig.fs = {
    readFileSync: path => fs.readFileSync(path) as Buffer,
  };
  fs.mkdirpSync(globalConfig.dir);

  const { mkdir, mkdirp, rmdir, unlink, writeFile } = fs;
  return {
    fs: { mkdir, mkdirp, rmdir, unlink, writeFile, join },
    createDotenv: (variation, pairs) => {
      fs.writeFileSync(
        join(globalConfig.dir, `.env.${variation}`),
        pairs.join("\n")
      );
    },
    expectToMatchSnapshot: filename => {
      expect(
        fs.readFileSync(join(outputPath, filename), "utf-8")
      ).toMatchSnapshot(filename);
    },
  };
}

describe("webpack-html-powerstrip", () => {
  it("generates multiple htmls", done => {
    const { fs, createDotenv, expectToMatchSnapshot } = setup();
    createDotenv("qa", [
      "REACT_APP_NODE_ENV=qa",
      'MESSAGE="This is QA"',
      'REACT_APP_SECRET="foobarbaz"',
    ]);
    createDotenv("staging", [
      "REACT_APP_NODE_ENV=stating",
      'MESSAGE="This is Staging"',
      'REACT_APP_SECRET="barbazfoo"',
    ]);
    createDotenv("production", [
      "REACT_APP_NODE_ENV=production",
      'MESSAGE="This is Production"',
      'REACT_APP_SECRET="bazfoobar"',
    ]);
    const config: webpack.Configuration = {
      entry: {
        main: join(__dirname, "__fixtures__", "entry.js"),
      },
      output: {
        path: outputPath,
        filename: "[name].js",
      },
      plugins: [new HtmlWebpackPlugin()],
    };
    const compiler = webpack(
      rewire(config, "production", {
        variations: ["qa", "staging", "production"],
      }),
      (err, _) => {
        expect(err).toBeNull();
        expectToMatchSnapshot("index.qa.html");
        expectToMatchSnapshot("index.staging.html");
        expectToMatchSnapshot("index.production.html");
        done();
      }
    ) as webpack.Compiler;
    compiler.outputFileSystem = fs;
  });
});
