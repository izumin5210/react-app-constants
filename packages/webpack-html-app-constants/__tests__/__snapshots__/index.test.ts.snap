// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack-html-powerstrip generates multiple htmls: index.production.html 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"UTF-8\\">
    <title>Webpack App</title>
  <meta name=\\"app-constants\\" content=\\"eyJOT0RFX0VOViI6InByb2R1Y3Rpb24iLCJTRUNSRVQiOiJiYXpmb29iYXIifQ==\\"></head>
  <body>
  <script type=\\"text/javascript\\" src=\\"main.js\\"></script></body>
</html>"
`;

exports[`webpack-html-powerstrip generates multiple htmls: index.qa.html 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"UTF-8\\">
    <title>Webpack App</title>
  <meta name=\\"app-constants\\" content=\\"eyJOT0RFX0VOViI6InFhIiwiU0VDUkVUIjoiZm9vYmFyYmF6In0=\\"></head>
  <body>
  <script type=\\"text/javascript\\" src=\\"main.js\\"></script></body>
</html>"
`;

exports[`webpack-html-powerstrip generates multiple htmls: index.staging.html 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"UTF-8\\">
    <title>Webpack App</title>
  <meta name=\\"app-constants\\" content=\\"eyJOT0RFX0VOViI6InN0YXRpbmciLCJTRUNSRVQiOiJiYXJiYXpmb28ifQ==\\"></head>
  <body>
  <script type=\\"text/javascript\\" src=\\"main.js\\"></script></body>
</html>"
`;
